<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [179]:
import pandas as pd
import psycopg2

In [180]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [181]:
connection = psycopg2.connect(
    dbname =
    user =
    password =
    host =
    port =
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [182]:
# текст запроса
query_3_1 = f'''
/* 
  Этот запрос считает общее количество записей в таблице 'vacancies'.
*/
select
    count(*) count_vacancies
from
    public.vacancies
'''

In [183]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [184]:
# текст запроса
query_3_2 = f'''
/* 
  Этот запрос считает общее количество записей в таблице 'employers'.
*/
select
    count(*) count_employers
from
    public.employers
'''

In [185]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count_employers
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [186]:
# текст запроса
query_3_3 = f'''
/* 
  Этот запрос считает общее количество записей в таблице 'areas'.
*/
select
    count(*) count_areas
from
    public.areas
'''

In [187]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count_areas
0         1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [188]:
# текст запроса
query_3_4 = f'''
/* 
  Этот запрос считает общее количество записей в таблице 'industries'.
*/
select
    count(*) count_industries
from
    public.industries
'''

In [189]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count_industries
0               294

***

Выводы по предварительному анализу данных:
1. Количество вакансий в базе данных составляет - 49197. Вакансий в базе в 2 раза больше, чем работадателей, т.е. в среднем 2 вакансии на 1 работадателя.
2. Количество работодателей в базе данных составляет - 23501. Работадателей в базе в 2 раза меньше, чем вакансий, т.е. в среднем 1 работадатель на 2 вакансии.
3. Количество регионов (городов) в базе данных составляет - 1362. Представлены практически все основные города России и ближнего зарубежья.
4. Количество сфер деятельности в базе данных составляет - 294.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [190]:
# текст запроса
query_4_1 = f'''
/* 
  Этот запрос выбирает названия регионов и количество вакансий в каждом регионе.
  Данные извлекаются из таблиц 'areas' и 'vacancies'.
  Запрос выполняет объединение таблиц по полю 'area_id', группирует данные по названию региона,
  сортирует их по количеству вакансий в порядке убывания и ограничивает результат первыми пятью записями.
*/
select
    a.name,
    count(v.area_id) cnt
from
    public.areas a
join
    public.vacancies v on v.area_id = a.id
group by a.name
order by 2 desc
limit 5
'''

In [191]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [192]:
# текст запроса
query_4_2 = f'''
/* 
  Этот запрос выбирает идентификаторы вакансий и количество непустых значений в полях 'salary_from' и 'salary_to'.
  Данные извлекаются из таблицы 'vacancies', которая находится в схеме 'public'.
  Запрос группирует данные по идентификатору вакансии, 
  и включает в результат только те записи, у которых есть хотя бы одно непустое значение в полях 'salary_from' или 'salary_to'.
*/
select
    v.id,
    count(v.salary_from) salary_from,
    count(v.salary_to) salary_to
from
    public.vacancies v
group by v.id
having count(v.salary_from) + count(v.salary_to) > 0
'''

In [193]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


id  salary_from  salary_to
0       6555939            1          0
1      14694668            1          0
2      14820007            1          0
3      16142474            1          1
4      17987111            1          0
...         ...          ...        ...
24068  66195906            1          1
24069  66196086            1          0
24070  66196283            1          1
24071  66196369            1          1
24072  66196438            1          1

[24073 rows x 3 columns]

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [194]:
# текст запроса
query_4_3 = f'''
/* 
  Этот запрос вычисляет средние значения полей 'salary_from' и 'salary_to' в таблице 'vacancies'.
  Таблица 'vacancies' находится в схеме 'public'.
  Результат округляется до 8 знаков после запятой.
*/
select
    cast(avg(salary_from) as decimal(8)) avg_salary_from,
    cast(avg(salary_to) as decimal(8)) avg_salary_to
from
    public.vacancies
'''


In [195]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [196]:
# текст запроса
query_4_4 = f'''
/* 
  Этот запрос выбирает комбинации графиков работы ('schedule') и типов занятости ('employment'),
  а также количество вакансий для каждой комбинации из таблицы 'vacancies'.
  Данные группируются по графику и типу занятости, сортируются по количеству вакансий в порядке убывания.
  Результат включает вторую по популярности комбинацию за счет использования 'offset 1' и 'limit 1'.
*/
select
    schedule,
    employment,
    count(*) vacancy_count
from
    public.vacancies
group by
    schedule,
    employment
order by
    vacancy_count desc
offset 1
limit 1
'''

In [197]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  vacancy_count
0  Удаленная работа  Полная занятость           7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [198]:
# текст запроса
query_4_5 = f'''
/* 
  Этот запрос группирует вакансии по уровню требуемого опыта ('experience') 
  и подсчитывает количество вакансий для каждого уровня опыта.
  Результат сортируется по количеству вакансий в порядке возрастания.
*/
select
    experience,
    count(*) vacancy_count
from
    public.vacancies
group by
    experience
order by
    vacancy_count asc
'''

In [199]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

Выводы по детальному анализу вакансий:
1. Наибольшее количество вакансий находится в крупных городах-миллионниках. Безусловным лидером является Москва - 5333 вакансий.
2. Почти у половины вакансии определён уровень зарплаты - 24073 из 49197. Зарплаты остальных вакансий видимо определяются по результатам собеседования.
3. Средние значения для нижней и верхней границы зарплатной вилки составляют: 71065 и 110537 руб. соответственно. Эти показатели выше средних зарплат в России и ближнем зарубежье.
4. "Удаленная работа и Полная занятость" - 7802 вакансий.
5. Наиболее востребованный опыт работы "От 1 года до 3 лет" - 26152 вакансий.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [200]:
# текст запроса
query_5_1 = f'''
/* 
  Этот запрос выбирает названия работодателей ('name') и подсчитывает количество вакансий,
  связанных с каждым из них, из таблиц 'vacancies' и 'employers'.
  Для связи вакансий с работодателями используется поле 'employer_id'.
  Результат группируется по названию работодателя, сортируется по количеству вакансий в порядке убывания.
  Выводятся только первые пять результатов за счет использования 'limit 5'.
*/
select
    e.name,
    count(v.employer_id) count_vacancies
from
    public.vacancies v
join
    public.employers e on e.id = v.employer_id
group by
    e.name
order by
    count_vacancies desc
limit 5
'''

In [201]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count_vacancies
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [202]:
# текст запроса
query_5_2 = f'''
/* 
  Этот запрос выбирает названия регионов ('name') из таблицы 'areas'.
  Затем он подсчитывает количество уникальных работодателей ('count_employers') и количество вакансий ('count_vacancies')
  в каждом регионе, где нет вакансий, используя LEFT JOIN для таблиц 'vacancies' и 'employers'.
  Результат группируется по названию региона, сортируется по количеству уникальных работодателей в порядке убывания.
  Выводится только первый регион с наибольшим количеством уникальных работодателей за счет использования 'limit 1'.
*/
select
    a.name region_name,
    count(DISTINCT e.id) count_employers,
    count(v.id) count_vacancies
from
    public.areas a
left join
    public.vacancies v on a.id = v.area_id
left join
    public.employers e on a.id = e.area
where
    v.id is null
group by
    a.name
order by
    count(DISTINCT e.id) desc
limit 1
'''

In [203]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region_name  count_employers  count_vacancies
0      Россия              410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [204]:
# текст запроса
query_5_3 = f'''
/* 
  Этот запрос выбирает названия работодателей ('name') и подсчитывает количество уникальных регионов ('area_id'),
  в которых у каждого работодателя есть вакансии, из таблиц 'vacancies' и 'employers'.
  Результат группируется по названию работодателя, сортируется по количеству уникальных регионов в порядке убывания.
  Выводится только первый результат за счет использования 'limit 1'.
*/
select
    e.name,
    count(DISTINCT area_id) unique_area
from
    public.vacancies v
join
    public.employers e on e.id = v.employer_id
group by
    e.name
order by
    count(DISTINCT area_id) desc
limit 1
'''

In [205]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  unique_area
0  Яндекс          181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [206]:
# текст запроса
query_5_4 = f'''
/* 
  Этот запрос подсчитывает количество уникальных работодателей ('employer_count'), 
  не связанных с какой-либо отраслью, из таблиц 'employers' и 'employers_industries'. 
  Используется LEFT JOIN для связи работодателей и отраслей.
  Результат фильтруется по отсутствию связи работодателя с отраслью (industry_id is null).
*/
select
    count(distinct e.id) employer_count
from
    public.employers e
left join
    public.employers_industries ei on e.id = ei.employer_id
where
    ei.industry_id is null
'''

In [207]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employer_count
0            8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [208]:
# текст запроса
query_5_5 = f'''
/* 
  Этот запрос выбирает название работодателя ('name'), который имеет связи с четырьмя отраслями,
  из таблиц 'employers' и 'employers_industries'.
  Используется JOIN для связи работодателей и отраслей.
  Результат группируется по идентификатору работодателя и его названию, фильтруется по количеству связей с отраслями равному 4.
  Затем результат сортируется по названию работодателя в алфавитном порядке и выводится третий по алфавиту с помощью 'limit 1 offset 2'.
*/
select
    e.name
from
    public.employers e
join
    public.employers_industries ei on e.id = ei.employer_id
group by
    e.id,
    e.name
having
    count(ei.industry_id) = 4
order by
    e.name
limit 1 offset 2
'''

In [209]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [210]:
# текст запроса
query_5_6 = f'''
/* 
  Этот запрос подсчитывает количество уникальных работодателей ('employer_count'), 
  связанных с отраслью 'Разработка программного обеспечения', 
  из таблиц 'employers', 'employers_industries' и 'industries'. 
  Используются JOIN для связи работодателей, отраслей и их связей.
  Результат фильтруется по названию отрасли 'Разработка программного обеспечения'.
*/
select
    count(distinct e.id) employer_count
from
    public.employers e
join
    public.employers_industries ei on e.id = ei.employer_id
join
    public.industries i on ei.industry_id = i.id
where
    i.name = 'Разработка программного обеспечения'
'''

In [211]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


employer_count
0            3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [212]:
# код для получения списка городов-милионников
city = ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Нижний Новгород',
        'Казань', 'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск',
        'Пермь', 'Воронеж', 'Волгоград', 'Краснодар')

In [213]:
# текст запроса
query_5_7 = f'''
/* Определяем количество вакансий, связанных с Яндексом, в указанных городах */
with yandex_vacancies as (
    select
        a.name region_name,
        count(v.id) cnt
    from
        public.vacancies v
    join
        public.areas a on v.area_id = a.id
    join
        public.employers e on v.employer_id = e.id
    where
        e.name = 'Яндекс'
        and a.name in {city}
    group by
        a.name
)
/* Сводим результаты по регионам и вычисляем общее количество вакансий */
select
    region_name,
    cnt
from
    yandex_vacancies
    
union all

select
    'Total',
    sum(cnt)::int4
from
    yandex_vacancies
order by
    cnt desc
    
'''

In [214]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


region_name  cnt
0             Total  485
1            Москва   54
2   Санкт-Петербург   42
3      Екатеринбург   39
4   Нижний Новгород   36
5       Новосибирск   35
6           Воронеж   32
7         Краснодар   30
8            Самара   26
9               Уфа   26
10   Ростов-на-Дону   25
11            Пермь   25
12           Казань   25
13        Волгоград   24
14       Красноярск   23
15        Челябинск   22
16             Омск   21

***

Выводы по анализу работодателей:
1. Наибольшее число вакансий среди работодателей принадлежит "Яндексу" - 1933, это большой отрыв от находящегося на втором месте "Ростелекома" - 491.
2. Среди регионов наибольшее число вакансий у России (410).
3. "Яндекс" как и ожидалось, обладают наибольшим числом регионов с вакансиями 181.
4. Довольно много работодателей не указало сферу своей деятельности 8419 из 23175.
5. У многих компаний указано несколько сфер деятельности (четыре сферы деятельности более чем у 1000 компаний).
6. Более 3,5 тысяч компаний указали сферу деятельности "Разработка программного обеспечения".
7. Наибольшее количество вакансий в крупных городах. Так по "Яндексу" в Москве - 54 вакансии, в Санкт-Петербурге - 42 вакансии.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [215]:
# текст запроса
query_6_1 = f'''
/* Подсчитываем количество вакансий с ключевыми словами "data" или "данные" в названии */
select
    count(v.name) data_vacancies
from
    public.vacancies v
where
    lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [216]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


data_vacancies
0            1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [217]:
# текст запроса
query_6_2 = f'''
/* Подсчитываем количество вакансий для начинающих в области Data Science и Machine Learning */
select
    count(*) junior_vacancies
from
    public.vacancies v
where
    (
        lower(v.name) like '%data scientist%' or
        lower(v.name) like '%data science%' or
        lower(v.name) like '%исследователь данных%' or
        (lower(v.name) like '%ml%' AND lower(v.name) NOT LIKE '%html%') or
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%'
    ) and
    (
        lower(v.name) like '%junior%' or
        lower(v.experience) like '%нет опыта%' or
        lower(v.employment) like '%стажировка%'
    )
'''

In [218]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


junior_vacancies
0                51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [219]:
# текст запроса
query_6_3 = f'''
/* 
   Подсчитываем количество вакансий в области Data Science с требуемыми навыками в SQL и PostgreSQL.
*/
select
    count(*) ds_vacancies
from
    public.vacancies v
where
    (
        lower(v.name) like '%data scientist%' or
        lower(v.name) like '%data science%' or
        lower(v.name) like '%исследователь данных%' or
        (lower(v.name) like '%ml%' AND lower(v.name) NOT LIKE '%html%') or
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%'
    )
    and
    (
        v.key_skills like '%SQL%' or
        v.key_skills like '%postgres%'
    )
'''

In [220]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


ds_vacancies
0           229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [221]:
# текст запроса
query_6_4 = f'''
/* 
   Подсчитываем количество вакансий в области Data Science, требующих навыков программирования на Python.
*/
select
    count(*) python_DS_vacancies
from
    public.vacancies v
where
    (
        lower(v.name) like '%data scientist%' or
        lower(v.name) like '%data science%' or
        lower(v.name) like '%исследователь данных%' or
        (lower(v.name) like '%ml%' AND lower(v.name) NOT LIKE '%html%') or
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%'
    )
    and
    (
        v.key_skills like '%Python%'
    )
'''

In [222]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


python_ds_vacancies
0                  357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [223]:
# текст запроса
query_6_5 = f'''
/* 
   Вычисляем среднее количество навыков, указанных в описании вакансии, 
   для вакансий в области Data Science.
*/
select
    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), ''))+1), 2) as avg_skills
from
	public.vacancies
where
	(name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%'
    and name not like '%HTML%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
'''

In [224]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


avg_skills
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [225]:
# текст запроса
query_6_6 = f'''
/*
   Этот запрос выбирает уникальные уровни опыта ('experience') 
   и вычисляет среднюю зарплату ('mean_salary') для вакансий в области Data Science и Machine Learning.
   Средняя зарплата вычисляется как среднее арифметическое между salary_from и salary_to, если оба значения не являются NULL,
   или как значение salary_from или salary_to, если одно из них равно NULL.
   Затем результат группируется по уровню опыта.
*/
select distinct
    experience,
    round(avg(coalesce((salary_from + salary_to)/2, salary_from, salary_to)), 0) as mean_salary
from
    public.vacancies
where
    (name ilike '%data scientist%'
    or name ilike '%data science%'
    or name ilike '%исследователь данных%'
    or (name like '%ML%'
    and name not like '%HTML%')
    or name ilike '%machine learning%'
    or name ilike '%машинн%обучен%')
    and (salary_from is not null
    or salary_to is not null)
group by 1
'''

In [226]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0           Нет опыта      74643.0
1  От 1 года до 3 лет     139675.0
2       От 3 до 6 лет     243115.0

***

Выводы по предметному анализу:
1. Имеет отношение к данным 1771 вакансия.
2. Для начинающего дата-сайентиста имеется 51 подходящия вакансия.
3. Имеется 229 вакансия для DS, в которых в качестве ключевого навыка указан SQL или postgres.
4. Python довольно популярен в требованиях работодателей к DS. Количество вакансий, в которых в качестве ключевого навыка указан Python - 357.
5. В среднем указывают в вакансиях для DS 6 ключевых навыков.
6. Для специалистов DS в среднем указывают довольно высокую зарплату. Чем выше опыт, тем более высокая зарплата. Для опыта работы от 3 до 6 лет в среднем около 243 тысяч руб.

# Общий вывод по проекту

Подведем итог исследования, обобщим выводы.
1. Всего вакансий - 49197. Вакансий, имеющих отношение к данным - 1771. Больше всего вакансий: Системный администратор (1761), Программист 1С (1153), Бизнес-аналитик (488), Специалист технической поддержки (457) и Системный аналитик (432).
2. На основании анализа данных можно сделать заключение, что специалисты в IT-отрасли очень востребованы на рынке труда.
3. Было бы интересно провести анализ не только вакансий, но и резюме соискателей. Сравнить предложения вакансий работодателей и запросов соискателей.

In [227]:
# Определим каких вакансий больше всего
query_7_1 = f'''
select
    v.name,
    count(v.id) as count_vac
from
    public.vacancies as v
group by 1
order by 2 desc
            '''

In [228]:
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Koop_\AppData\Local\Temp\ipykernel_12160\2931715827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


name  count_vac
0                     Системный администратор       1761
1                              Программист 1С       1153
2                             Бизнес-аналитик        488
3            Специалист технической поддержки        457
4                          Системный аналитик        432
...                                       ...        ...
21218            Генеральный менеджер Проекта          1
21219    Ведущий инженер-программист (1С:УПП)          1
21220       Главный специалист IT направления          1
21221  Сотрудник Отдела информационных систем          1
21222             Старший финансовый менеджер          1

[21223 rows x 2 columns]

## завершение исследования

In [229]:
# не забываем закрыть соединение после окончания работы
connection.close()